src/network/protocol/messages/addr.zig

```diff
@@ -0,0 +1,180 @@
+        }
+        //try CompactSizeUint.new(@intCast(self.ip_addresses.len)).encodeToWriter(w);
+        try CompactSizeUint.new(self.ip_addresses.len).encodeToWriter(w);
+
+        // Serialize each IP address
+        for (self.ip_addresses) |*ip_address| {
+            try w.writeInt(u32, ip_address.time, .little);
+            try w.writeInt(u64, ip_address.services, .little);
+        try w.writeAll(std.mem.asBytes(&ip_address.ip));
+            try w.writeInt(u16, ip_address.port, .big);
+        }
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+    ///
+    /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const AddrMessage, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        const writer = fbs.writer();
+        try self.serializeToWriter(writer);
+    }
+
+    /// Serialize a message as bytes and return them.
+    pub fn serialize(self: *const AddrMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    /// Deserialize a Reader bytes as a `AddrMessage`
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !AddrMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects r to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects r to have fn 'readNoEof'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readAll")) @compileError("Expects r to have fn 'readAll'.");
+        }
+
+        var vm: AddrMessage = undefined;
+        const ip_address_count = try CompactSizeUint.decodeReader(r);
+
+        // Allocate space for IP addresses
+        const ip_addresses = try allocator.alloc(NetworkIPAddr, ip_address_count.value());
+
+        vm.ip_addresses = ip_addresses;
+
+        for (vm.ip_addresses) |*ip_address| {
+            ip_address.time = try r.readInt(u32, .little);
+            ip_address.services = try r.readInt(u64, .little);
+            try r.readNoEof(&ip_address.ip);
+            ip_address.port = try r.readInt(u16, .big);
+        }
+
+        return vm;
+    }
+
+    /// Deserialize bytes into a `AddrMessage`
+    pub fn deserializeSlice(allocator: std.mem.Allocator, bytes: []const u8) !AddrMessage {
+        var fbs = std.io.fixedBufferStream(bytes);
+        const reader = fbs.reader();
+        return try AddrMessage.deserializeReader(allocator, reader);
+    }
+
+    pub fn hintSerializedLen(self: AddrMessage) usize {
+        // 4 + 8 + 16 + 2
+        const fixed_length_per_ip = 30;
+        return 1 +  self.ip_addresses.len * fixed_length_per_ip;// 1 for CompactSizeUint
```

Comment:
CompactSizeUint len is not 1, it's CompactSizeUint.new(ip_addresses.len).hint_encoded_size()

The bigger the number, the more bytes it takes to write it